File: databases/cockroachdb-patterns/assets/retry-transaction.py
#!/usr/bin/env python3
"""
retry-transaction.py

Demonstrates proper CockroachDB transaction retry logic in Python using psycopg2.
Handles SQLSTATE 40001 (serialization failure) with exponential backoff and jitter.

Usage:
    pip install psycopg2-binary
    python retry-transaction.py

Environment:
    DATABASE_URL  - Connection string (default: postgresql://root@localhost:26257/appdb)
"""

import os
import time
import random
import logging
import psycopg2
import psycopg2.errors
from contextlib import contextmanager
from decimal import Decimal
from typing import Callable, TypeVar, Optional

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
logger = logging.getLogger(__name__)

T = TypeVar("T")

DEFAULT_DSN = "postgresql://root@localhost:26257/appdb?sslmode=disable"
MAX_RETRIES = 10
BASE_DELAY = 0.01  # 10ms
MAX_DELAY = 5.0    # 5 seconds


def execute_with_retry(
    conn,
    fn: Callable,
    max_retries: int = MAX_RETRIES,
    base_delay: float = BASE_DELAY,
    max_delay: float = MAX_DELAY,
) -> Optional[T]:
    """
    Execute a database operation with automatic retry on serialization failure.

    Uses the SAVEPOINT-based retry protocol recommended by CockroachDB:
      BEGIN; SAVEPOINT cockroach_restart; <ops>; RELEASE SAVEPOINT; COMMIT;
      On 40001: ROLLBACK TO SAVEPOINT cockroach_restart; retry

    Args:
        conn: psycopg2 connection (autocommit must be False)
        fn: Callable that accepts a cursor and performs DB operations
        max_retries: Maximum number of retry attempts
        base_delay: Initial delay between retries (seconds)
        max_delay: Maximum delay between retries (seconds)

    Returns:
        The return value of fn, if any.

    Raises:
        Exception: If max retries exceeded or non-retryable error occurs.
    """
    conn.autocommit = False

    for attempt in range(max_retries):
        try:
            with conn.cursor() as cur:
                cur.execute("SAVEPOINT cockroach_restart")
                result = fn(cur)
                cur.execute("RELEASE SAVEPOINT cockroach_restart")
            conn.commit()
            if attempt > 0:
                logger.info("Transaction succeeded after %d retries", attempt)
            return result

        except psycopg2.errors.SerializationFailure as e:
            # Retryable: rollback to savepoint and retry
            conn.rollback()
            delay = _calculate_backoff(attempt, base_delay, max_delay)
            logger.warning(
                "Retry %d/%d: serialization failure (40001), waiting %.3fs",
                attempt + 1, max_retries, delay,
            )
            time.sleep(delay)

        except psycopg2.Error as e:
            # Non-retryable database error
            conn.rollback()
            raise RuntimeError(f"Non-retryable database error: {e}") from e

    raise RuntimeError(f"Transaction failed after {max_retries} retries")


def _calculate_backoff(attempt: int, base_delay: float, max_delay: float) -> float:
    """Calculate exponential backoff with jitter."""
    backoff = base_delay * (2 ** attempt)
    jitter = random.uniform(0, backoff * 0.5)
    return min(backoff + jitter, max_delay)


# ---------------------------------------------------------------------------
# Example: Account Transfer
# ---------------------------------------------------------------------------

def setup_schema(conn):
    """Create tables and seed data for the demo."""
    conn.autocommit = True
    with conn.cursor() as cur:
        cur.execute("""
            CREATE TABLE IF NOT EXISTS accounts (
                id STRING PRIMARY KEY,
                name STRING NOT NULL,
                balance DECIMAL(12,2) NOT NULL DEFAULT 0,
                updated_at TIMESTAMPTZ NOT NULL DEFAULT now(),
                CHECK (balance >= 0)
            );

            CREATE TABLE IF NOT EXISTS transfers (
                id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                from_account STRING NOT NULL REFERENCES accounts(id),
                to_account STRING NOT NULL REFERENCES accounts(id),
                amount DECIMAL(12,2) NOT NULL CHECK (amount > 0),
                created_at TIMESTAMPTZ NOT NULL DEFAULT now()
            );

            UPSERT INTO accounts (id, name, balance) VALUES
                ('account-a', 'Alice', 1000.00),
                ('account-b', 'Bob', 500.00);
        """)
    conn.autocommit = False
    logger.info("Schema and seed data ready.")


def transfer(cur, from_id: str, to_id: str, amount: Decimal):
    """Transfer funds between accounts within a transaction."""
    # Lock source account early with SELECT FOR UPDATE
    cur.execute(
        "SELECT balance FROM accounts WHERE id = %s FOR UPDATE",
        (from_id,),
    )
    row = cur.fetchone()
    if row is None:
        raise ValueError(f"Account {from_id} not found")

    balance = row[0]
    if balance < amount:
        raise ValueError(
            f"Insufficient funds: have {balance}, need {amount}"
        )

    # Debit source
    cur.execute(
        "UPDATE accounts SET balance = balance - %s, updated_at = now() WHERE id = %s",
        (amount, from_id),
    )

    # Credit destination
    cur.execute(
        "UPDATE accounts SET balance = balance + %s, updated_at = now() WHERE id = %s",
        (amount, to_id),
    )

    # Record transfer
    cur.execute(
        """INSERT INTO transfers (from_account, to_account, amount)
           VALUES (%s, %s, %s)""",
        (from_id, to_id, amount),
    )

    logger.info("Transferred $%s from %s to %s", amount, from_id, to_id)


def print_balances(conn):
    """Print current account balances."""
    with conn.cursor() as cur:
        cur.execute("SELECT id, name, balance FROM accounts ORDER BY id")
        rows = cur.fetchall()

    print("\nAccount Balances:")
    print("-" * 50)
    for account_id, name, balance in rows:
        print(f"  {account_id:<15} {name:<15} ${balance:>10.2f}")
    print("-" * 50)


def main():
    dsn = os.getenv("DATABASE_URL", DEFAULT_DSN)
    logger.info("Connecting to %s", dsn)

    conn = psycopg2.connect(dsn)
    try:
        setup_schema(conn)

        # Execute transfer with automatic retry on serialization failure
        execute_with_retry(
            conn,
            lambda cur: transfer(cur, "account-a", "account-b", Decimal("50.00")),
        )

        print_balances(conn)
        logger.info("Transfer completed successfully!")

    finally:
        conn.close()


if __name__ == "__main__":
    main()

File: databases/cockroachdb-patterns/assets/retry-transaction.go
// retry-transaction.go
//
// Demonstrates proper CockroachDB transaction retry logic in Go using pgx.
// Handles SQLSTATE 40001 (serialization failure) with exponential backoff.
//
// Usage:
//   go run retry-transaction.go
//
// Requires:
//   go get github.com/jackc/pgx/v5
//   go get github.com/jackc/pgx/v5/pgxpool

package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
)

const (
	defaultDSN    = "postgresql://root@localhost:26257/appdb?sslmode=disable"
	maxRetries    = 10
	baseDelay     = 10 * time.Millisecond
	maxDelay      = 5 * time.Second
	serializationFailureCode = "40001"
)

func main() {
	ctx := context.Background()

	dsn := os.Getenv("DATABASE_URL")
	if dsn == "" {
		dsn = defaultDSN
	}

	pool, err := pgxpool.New(ctx, dsn)
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer pool.Close()

	if err := setup(ctx, pool); err != nil {
		log.Fatalf("Setup failed: %v", err)
	}

	// Transfer $50 from account A to account B with retry logic
	err = ExecuteWithRetry(ctx, pool, func(ctx context.Context, tx pgx.Tx) error {
		return transfer(ctx, tx, "account-a", "account-b", 50.00)
	})
	if err != nil {
		log.Fatalf("Transfer failed: %v", err)
	}

	printBalances(ctx, pool)
	log.Println("Transfer completed successfully!")
}

// ExecuteWithRetry runs a transaction function with automatic retry on
// serialization failure (SQLSTATE 40001). Uses exponential backoff with jitter.
func ExecuteWithRetry(ctx context.Context, pool *pgxpool.Pool, fn func(ctx context.Context, tx pgx.Tx) error) error {
	for attempt := 0; attempt < maxRetries; attempt++ {
		err := pgx.BeginTxFunc(ctx, pool, pgx.TxOptions{
			IsoLevel: pgx.Serializable,
		}, func(tx pgx.Tx) error {
			return fn(ctx, tx)
		})

		if err == nil {
			if attempt > 0 {
				log.Printf("Transaction succeeded after %d retries", attempt)
			}
			return nil
		}

		// Check if the error is a serialization failure (retryable)
		var pgErr *pgconn.PgError
		if errors.As(err, &pgErr) && pgErr.Code == serializationFailureCode {
			delay := calculateBackoff(attempt)
			log.Printf("Retry %d/%d: serialization failure, waiting %v", attempt+1, maxRetries, delay)

			select {
			case <-ctx.Done():
				return fmt.Errorf("context cancelled during retry: %w", ctx.Err())
			case <-time.After(delay):
				continue
			}
		}

		// Non-retryable error
		return fmt.Errorf("transaction failed with non-retryable error: %w", err)
	}

	return fmt.Errorf("transaction failed after %d retries", maxRetries)
}

// calculateBackoff returns exponential backoff duration with jitter.
func calculateBackoff(attempt int) time.Duration {
	backoff := float64(baseDelay) * math.Pow(2, float64(attempt))
	jitter := rand.Float64() * backoff * 0.5 // 0-50% jitter
	delay := time.Duration(backoff + jitter)
	if delay > maxDelay {
		delay = maxDelay
	}
	return delay
}

// transfer moves amount from one account to another within a transaction.
func transfer(ctx context.Context, tx pgx.Tx, fromID, toID string, amount float64) error {
	// Check source balance using SELECT FOR UPDATE to acquire lock early
	var balance float64
	err := tx.QueryRow(ctx,
		"SELECT balance FROM accounts WHERE id = $1 FOR UPDATE",
		fromID,
	).Scan(&balance)
	if err != nil {
		return fmt.Errorf("failed to read source balance: %w", err)
	}

	if balance < amount {
		return fmt.Errorf("insufficient funds: have %.2f, need %.2f", balance, amount)
	}

	// Debit source
	_, err = tx.Exec(ctx,
		"UPDATE accounts SET balance = balance - $1, updated_at = now() WHERE id = $2",
		amount, fromID,
	)
	if err != nil {
		return fmt.Errorf("failed to debit source: %w", err)
	}

	// Credit destination
	_, err = tx.Exec(ctx,
		"UPDATE accounts SET balance = balance + $1, updated_at = now() WHERE id = $2",
		amount, toID,
	)
	if err != nil {
		return fmt.Errorf("failed to credit destination: %w", err)
	}

	// Record the transfer
	_, err = tx.Exec(ctx, `
		INSERT INTO transfers (id, from_account, to_account, amount, created_at)
		VALUES (gen_random_uuid(), $1, $2, $3, now())`,
		fromID, toID, amount,
	)
	if err != nil {
		return fmt.Errorf("failed to record transfer: %w", err)
	}

	return nil
}

// setup creates the schema and seed data for the demo.
func setup(ctx context.Context, pool *pgxpool.Pool) error {
	_, err := pool.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS accounts (
			id STRING PRIMARY KEY,
			name STRING NOT NULL,
			balance DECIMAL(12,2) NOT NULL DEFAULT 0,
			updated_at TIMESTAMPTZ NOT NULL DEFAULT now(),
			CHECK (balance >= 0)
		);

		CREATE TABLE IF NOT EXISTS transfers (
			id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
			from_account STRING NOT NULL REFERENCES accounts(id),
			to_account STRING NOT NULL REFERENCES accounts(id),
			amount DECIMAL(12,2) NOT NULL CHECK (amount > 0),
			created_at TIMESTAMPTZ NOT NULL DEFAULT now()
		);

		UPSERT INTO accounts (id, name, balance) VALUES
			('account-a', 'Alice', 1000.00),
			('account-b', 'Bob', 500.00);
	`)
	return err
}

// printBalances prints current balances for all accounts.
func printBalances(ctx context.Context, pool *pgxpool.Pool) {
	rows, err := pool.Query(ctx, "SELECT id, name, balance FROM accounts ORDER BY id")
	if err != nil {
		log.Printf("Failed to query balances: %v", err)
		return
	}
	defer rows.Close()

	fmt.Println("\nAccount Balances:")
	fmt.Println("--------------------------------------------------")
	for rows.Next() {
		var id, name string
		var balance float64
		if err := rows.Scan(&id, &name, &balance); err != nil {
			log.Printf("Failed to scan row: %v", err)
			continue
		}
		fmt.Printf("  %-15s %-15s $%.2f\n", id, name, balance)
	}
	fmt.Println("--------------------------------------------------")
}
